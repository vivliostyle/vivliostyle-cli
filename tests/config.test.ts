import { expect, it, onTestFinished, vi } from 'vitest';
import { UseTemporaryServerRoot } from '../src/config/resolve.js';
import { VivliostyleConfigSchema } from '../src/config/schema.js';
import { getTaskConfig, maskConfig, resolveFixture } from './command-util.js';

const validConfigData = {
  title: 'title',
  author: 'author',
  theme: ['../themes/debug-theme'],
  entry: [
    'manuscript.md',
    {
      path: 'frontmatter.md',
      title: 'title',
      theme: {
        specifier: 'theme.css',
      },
    },
  ],
  entryContext: '.',
  output: [
    'output1.pdf',
    {
      path: 'output2.pdf',
      format: 'pdf',
    },
  ],
  size: 'size',
  pressReady: true,
  language: 'language',
  toc: {
    title: 'TOC',
    htmlPath: './toc.html',
    sectionDepth: 6,
  },
  cover: {
    src: './cover.png',
    name: 'Cover image alt',
    htmlPath: './mycover.html',
  },
  timeout: 1,
  workspaceDir: 'workspaceDir',
  vfm: {
    hardLineBreaks: true,
    disableFormatHtml: true,
  },
  readingProgression: 'rtl',
  browser: 'firefox',
  viewerParam: 'foo=bar',
  copyAsset: {
    includes: ['xx/yy', '**/zz'],
    excludes: ['*a*'],
    includeFileExtensions: ['zip'],
    excludeFileExtensions: ['png', 'jpg'],
  },
  base: '/root/vvv/',
  server: {
    host: true,
    port: 9876,
    proxy: {
      '/api': 'http://localhost:6789',
    },
  },
  static: {
    '/static': 'path/to/static',
    '/': ['root1', 'root2'],
  },
  temporaryFilePrefix: 'vvv.',
} satisfies VivliostyleConfigSchema;

it('parse vivliostyle config', async () => {
  const validConfig1 = await getTaskConfig(
    ['build'],
    resolveFixture('config'),
    validConfigData,
  );
  maskConfig(validConfig1);
  expect(validConfig1).toMatchSnapshot('config');
});

it('override option by CLI command', async () => {
  const config = await getTaskConfig(
    [
      'build',
      '-o',
      'yuno.pdf',
      '-o',
      'yuno',
      '-f',
      'webpub',
      '-T',
      'https://myTheme.example.com',
      '-s',
      'JIS-B5',
      '--title',
      'myTitle',
      '--author',
      'myAuthor',
      '--language',
      'myLanguage',
      '--timeout',
      '42',
      '--ignore-https-errors',
      '--executable-browser',
      'myFirefox',
      // '--browser',
      // 'webkit',
      '--style',
      'https://vivlostyle.org',
      '--user-style',
      './user/style/dummy.css',
      '--http',
      '--viewer',
      'https://vivliostyle.org/viewer/',
      '--viewer-param',
      'allowScripts=false&pixelRatio=16',
      '--proxy-server',
      'http://localhost:3128',
      '--proxy-bypass',
      '.example.com',
      '--proxy-user',
      'proxy-auth-user',
      '--proxy-pass',
      'proxy-auth-password',
      '--no-vite-config-file',
    ],
    resolveFixture('config'),
    validConfigData,
  );
  maskConfig(config);
  expect(config).toMatchSnapshot('config');
});

it('override option by environment variable', async () => {
  vi.stubEnv('HTTP_PROXY', 'https://localhost:9001');
  vi.stubEnv('NOPROXY', 'bypass.example.com');
  onTestFinished(() => {
    vi.unstubAllEnvs();
  });
  const validConfig1 = await getTaskConfig(
    ['build'],
    resolveFixture('config'),
    validConfigData,
  );
  expect(validConfig1.proxy?.server).toBe('https://localhost:9001');
  expect(validConfig1.proxy?.bypass).toBe('bypass.example.com');
});

it('deny invalid config', async () => {
  await expect(
    getTaskConfig(
      ['build'],
      resolveFixture('config'),
      // @ts-expect-error
      {
        output: {
          path: 'output',
          format: 'invalidFormat',
        },
      },
    ),
  ).rejects.toThrow();
});

it('deny config which has no entry', async () => {
  await expect(
    getTaskConfig(['build'], resolveFixture('config'), { entry: [] }),
  ).rejects.toThrow();
});

it('deny if any config file or input file is not set', async () => {
  await expect(
    getTaskConfig(['build'], resolveFixture('config')),
  ).rejects.toThrow();
});

it('deny if duplicate entry is set', async () => {
  await expect(
    getTaskConfig(['build'], resolveFixture('config'), {
      entry: ['index.md'],
      toc: true,
    }),
  ).rejects.toThrow();
});

it('yields a config with single markdown', async () => {
  const config = await getTaskConfig(
    ['build', 'sample.md'],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    entries: [
      {
        target:
          '__WORKSPACE__/tests/fixtures/config/__TEMPORARY_FILE_PREFIX__sample.html',
      },
    ],
    viewerInput: {
      type: 'webpub',
      manifestPath:
        '__WORKSPACE__/tests/fixtures/config/__TEMPORARY_FILE_PREFIX__publication.json',
      needToGenerateManifest: true,
    },
    exportAliases: expect.arrayContaining([
      {
        source:
          '__WORKSPACE__/tests/fixtures/config/__TEMPORARY_FILE_PREFIX__sample.html',
        target: '__WORKSPACE__/tests/fixtures/config/sample.html',
      },
      {
        source:
          '__WORKSPACE__/tests/fixtures/config/__TEMPORARY_FILE_PREFIX__publication.json',
        target: '__WORKSPACE__/tests/fixtures/config/publication.json',
      },
    ]),
  });
});

it('imports single html file', async () => {
  const config = await getTaskConfig(
    ['build', 'sample.html'],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'webbook',
      webbookEntryUrl: '/vivliostyle/sample.html',
    },
  });
});

it('yields a config with single input and vivliostyle config', async () => {
  const config = await getTaskConfig(
    ['build', 'nestedDir/01.md'],
    resolveFixture('config'),
    validConfigData,
  );
  maskConfig(config);
  expect(config).toMatchObject({
    entries: [
      {
        target: '__WORKSPACE__/tests/fixtures/config/nestedDir/vvv.01.html',
      },
    ],
    viewerInput: {
      type: 'webpub',
      manifestPath:
        '__WORKSPACE__/tests/fixtures/config/nestedDir/vvv.publication.json',
      needToGenerateManifest: true,
    },
  });
});

it('imports a EPUB file', async () => {
  const config = await getTaskConfig(
    ['build', 'adaptive.epub', '-o', 'epub.pdf'],
    resolveFixture('epubs'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'epub',
      epubPath: '__WORKSPACE__/tests/fixtures/epubs/adaptive.epub',
      epubTmpOutputDir:
        '__WORKSPACE__/tests/fixtures/epubs/__TEMPORARY_FILE_PREFIX__adaptive.epub',
    },
  });
});

it('imports a EPUB OPF file', async () => {
  const config = await getTaskConfig(
    ['build', 'adaptive/OPS/content.opf', '-o', 'epub-opf.pdf'],
    resolveFixture('epubs'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'epub-opf',
      epubOpfPath:
        '__WORKSPACE__/tests/fixtures/epubs/adaptive/OPS/content.opf',
    },
  });
});

it('imports a webbook compliant to W3C Web publication', async () => {
  const config = await getTaskConfig(
    ['build', 'w3c-webpub/publication.json', '-o', 'w3c-webpub-out'],
    resolveFixture('webbooks'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'webpub',
      manifestPath:
        '__WORKSPACE__/tests/fixtures/webbooks/w3c-webpub/publication.json',
      needToGenerateManifest: false,
    },
  });
});

it('imports a webbook compliant to Readium Web publication', async () => {
  const config = await getTaskConfig(
    ['build', 'readium-webpub/manifest.jsonld', '-o', 'readium-webpub-out'],
    resolveFixture('webbooks'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'webpub',
      manifestPath:
        '__WORKSPACE__/tests/fixtures/webbooks/readium-webpub/manifest.jsonld',
      needToGenerateManifest: false,
    },
  });
});

it('imports a https URL', async () => {
  const config = await getTaskConfig(
    [
      'build',
      'https://vivliostyle.github.io/vivliostyle_doc/ja/vivliostyle-user-group-vol1/',
    ],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'webbook',
      webbookEntryUrl:
        'https://vivliostyle.github.io/vivliostyle_doc/ja/vivliostyle-user-group-vol1/',
    },
  });
});

it('imports a HTML file with file protocol', async () => {
  const config = await getTaskConfig(
    ['build', `file://${resolveFixture('config')}/sample.html`],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    viewerInput: {
      type: 'webbook',
      webbookEntryUrl: '/vivliostyle/sample.html',
    },
  });
});

it('yields a config from frontmatter', async () => {
  const config = await getTaskConfig(
    ['build', 'frontmatter.md'],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config.entries[0].title).toBe('Frontmatter');
});

it('allow a loose specifier of a theme direcory', async () => {
  const validConfig = await getTaskConfig(['build'], resolveFixture('config'), {
    entry: 'manuscript.md',
    output: 'output1.pdf',
    theme: 'themes/foo',
  });
  maskConfig(validConfig);
  expect(validConfig).toMatchObject({
    entries: [
      {
        themes: [
          {
            type: 'package',
            name: 'foo',
            specifier: '__WORKSPACE__/tests/fixtures/config/themes/foo',
            location:
              '__WORKSPACE__/tests/fixtures/config/.vivliostyle/themes/node_modules/foo',
          },
        ],
      },
    ],
  });
});

it('resolves server root dir for single input file', async () => {
  const config = await getTaskConfig(
    ['build', 'nestedDir/01.md'],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    serverRootDir: '__WORKSPACE__/tests/fixtures/config/nestedDir',
    workspaceDir: '__WORKSPACE__/tests/fixtures/config/nestedDir',
  });
});

it('locates server root dir for EPUB OPF file', async () => {
  const config = await getTaskConfig(
    ['build', 'adaptive/OPS/content.opf'],
    resolveFixture('epubs'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    serverRootDir: '__WORKSPACE__/tests/fixtures/epubs/adaptive',
    workspaceDir: '__WORKSPACE__/tests/fixtures/epubs/adaptive',
  });
});

it('uses temporary dir for server root dir', async () => {
  const config = await getTaskConfig(
    ['build', 'https://example.com'],
    resolveFixture('config'),
  );
  maskConfig(config);
  expect(config).toMatchObject({
    serverRootDir: UseTemporaryServerRoot,
    workspaceDir: '__WORKSPACE__/tests/fixtures/config',
  });
});

it('deny config which has incompatible image', async () => {
  await expect(
    getTaskConfig(['build'], resolveFixture('config'), {
      ...validConfigData,
      image: 'ghcr.io/vivliostyle/cli:0.0.0',
    }),
  ).rejects.toThrow();
});
