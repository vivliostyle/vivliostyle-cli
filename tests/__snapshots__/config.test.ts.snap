// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`override option by CLI command > config 1`] = `
Object {
  "author": "myAuthor",
  "base": "/root/vvv",
  "bleed": undefined,
  "browser": Object {
    "executablePath": "myFirefox",
    "type": "firefox",
  },
  "copyAsset": Object {
    "excludes": Array [
      "*a*",
    ],
    "fileExtensions": Array [
      "jpeg",
      "svg",
      "gif",
      "webp",
      "apng",
      "ttf",
      "otf",
      "woff",
      "woff2",
      "zip",
    ],
    "includes": Array [
      "xx/yy",
      "**/zz",
    ],
  },
  "cover": Object {
    "name": "Cover image alt",
    "src": "__WORKSPACE__/tests/fixtures/config/cover.png",
  },
  "cropMarks": false,
  "cropOffset": undefined,
  "css": undefined,
  "customStyle": "https://vivlostyle.org",
  "customUserStyle": "theme.css",
  "documentProcessorFactory": [Function],
  "entries": Array [
    Object {
      "coverImageAlt": "Cover image alt",
      "coverImageSrc": "__WORKSPACE__/tests/fixtures/config/cover.png",
      "rel": "cover",
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/mycover.html",
      "themes": Array [],
      "title": "myTitle",
    },
    Object {
      "rel": "contents",
      "sectionDepth": 6,
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/toc.html",
      "themes": Array [
        Object {
          "location": "https://myTheme.example.com",
          "name": "myTheme.example.com",
          "type": "uri",
        },
      ],
      "tocTitle": "TOC",
      "transform": Object {
        "transformDocumentList": undefined,
        "transformSectionList": undefined,
      },
    },
    Object {
      "contentType": "text/markdown",
      "source": Object {
        "contentType": "text/markdown",
        "pathname": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
        "type": "file",
      },
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/manuscript.html",
      "themes": Array [
        Object {
          "location": "https://myTheme.example.com",
          "name": "myTheme.example.com",
          "type": "uri",
        },
      ],
      "title": "myTitle",
    },
    Object {
      "contentType": "text/markdown",
      "source": Object {
        "contentType": "text/markdown",
        "pathname": "__WORKSPACE__/tests/fixtures/config/frontmatter.md",
        "type": "file",
      },
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/frontmatter.html",
      "themes": Array [
        Object {
          "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/theme.css",
          "name": "theme.css",
          "source": "__WORKSPACE__/tests/fixtures/config/theme.css",
          "type": "file",
        },
      ],
      "title": "title",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "exportAliases": Array [],
  "ignoreHttpsErrors": true,
  "image": "__IMAGE__",
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/workspaceDir/publication.json",
    "format": "pub-manifest",
  },
  "language": "myLanguage",
  "logLevel": "info",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/yuno.pdf",
      "preflight": "press-ready",
      "preflightOption": Array [],
      "renderMode": "local",
    },
    Object {
      "format": "webpub",
      "path": "__WORKSPACE__/tests/fixtures/config/yuno",
    },
  ],
  "proxy": Object {
    "bypass": ".example.com",
    "password": "proxy-auth-password",
    "server": "http://localhost:3128",
    "username": "proxy-auth-user",
  },
  "quick": false,
  "readingProgression": "rtl",
  "rootUrl": "http://0.0.0.0:9876",
  "sandbox": false,
  "server": Object {
    "allowedHosts": Array [],
    "host": true,
    "port": 9876,
    "proxy": Object {
      "/api": "http://localhost:6789",
    },
  },
  "serverRootDir": "__WORKSPACE__/tests/fixtures/config",
  "singleDoc": false,
  "size": Object {
    "format": "JIS-B5",
  },
  "static": Object {
    "/": Array [
      "root1",
      "root2",
    ],
    "/static": Array [
      "path/to/static",
    ],
  },
  "temporaryFilePrefix": "__TEMPORARY_FILE_PREFIX__",
  "themeIndexes": Set {
    Object {
      "location": "https://myTheme.example.com",
      "name": "myTheme.example.com",
      "type": "uri",
    },
    Object {
      "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/theme.css",
      "name": "theme.css",
      "source": "__WORKSPACE__/tests/fixtures/config/theme.css",
      "type": "file",
    },
  },
  "themesDir": "__WORKSPACE__/tests/fixtures/config/workspaceDir/themes",
  "timeout": 42000,
  "title": "myTitle",
  "vfmOptions": Object {
    "disableFormatHtml": true,
    "hardLineBreaks": true,
  },
  "viewer": "https://vivliostyle.org/viewer/",
  "viewerInput": Object {
    "manifestPath": "__WORKSPACE__/tests/fixtures/config/workspaceDir/publication.json",
    "needToGenerateManifest": true,
    "type": "webpub",
  },
  "viewerParam": "allowScripts=false&pixelRatio=16",
  "viteConfig": undefined,
  "viteConfigFile": false,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config/workspaceDir",
}
`;

exports[`parse vivliostyle config > config 1`] = `
Object {
  "author": "author",
  "base": "/root/vvv",
  "bleed": undefined,
  "browser": Object {
    "executablePath": undefined,
    "type": "firefox",
  },
  "copyAsset": Object {
    "excludes": Array [
      "*a*",
    ],
    "fileExtensions": Array [
      "jpeg",
      "svg",
      "gif",
      "webp",
      "apng",
      "ttf",
      "otf",
      "woff",
      "woff2",
      "zip",
    ],
    "includes": Array [
      "xx/yy",
      "**/zz",
    ],
  },
  "cover": Object {
    "name": "Cover image alt",
    "src": "__WORKSPACE__/tests/fixtures/config/cover.png",
  },
  "cropMarks": false,
  "cropOffset": undefined,
  "css": undefined,
  "customStyle": undefined,
  "customUserStyle": undefined,
  "documentProcessorFactory": [Function],
  "entries": Array [
    Object {
      "coverImageAlt": "Cover image alt",
      "coverImageSrc": "__WORKSPACE__/tests/fixtures/config/cover.png",
      "rel": "cover",
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/mycover.html",
      "themes": Array [],
      "title": "title",
    },
    Object {
      "rel": "contents",
      "sectionDepth": 6,
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/toc.html",
      "themes": Array [
        Object {
          "importPath": undefined,
          "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/themes/node_modules/debug-theme",
          "name": "debug-theme",
          "registry": false,
          "specifier": "__WORKSPACE__/tests/fixtures/themes/debug-theme",
          "type": "package",
        },
      ],
      "tocTitle": "TOC",
      "transform": Object {
        "transformDocumentList": undefined,
        "transformSectionList": undefined,
      },
    },
    Object {
      "contentType": "text/markdown",
      "source": Object {
        "contentType": "text/markdown",
        "pathname": "__WORKSPACE__/tests/fixtures/config/manuscript.md",
        "type": "file",
      },
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/manuscript.html",
      "themes": Array [
        Object {
          "importPath": undefined,
          "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/themes/node_modules/debug-theme",
          "name": "debug-theme",
          "registry": false,
          "specifier": "__WORKSPACE__/tests/fixtures/themes/debug-theme",
          "type": "package",
        },
      ],
      "title": "title",
    },
    Object {
      "contentType": "text/markdown",
      "source": Object {
        "contentType": "text/markdown",
        "pathname": "__WORKSPACE__/tests/fixtures/config/frontmatter.md",
        "type": "file",
      },
      "target": "__WORKSPACE__/tests/fixtures/config/workspaceDir/frontmatter.html",
      "themes": Array [
        Object {
          "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/theme.css",
          "name": "theme.css",
          "source": "__WORKSPACE__/tests/fixtures/config/theme.css",
          "type": "file",
        },
      ],
      "title": "title",
    },
  ],
  "entryContextDir": "__WORKSPACE__/tests/fixtures/config",
  "exportAliases": Array [],
  "ignoreHttpsErrors": false,
  "image": "__IMAGE__",
  "input": Object {
    "entry": "__WORKSPACE__/tests/fixtures/config/workspaceDir/publication.json",
    "format": "pub-manifest",
  },
  "language": "language",
  "logLevel": "info",
  "outputs": Array [
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output1.pdf",
      "preflight": "press-ready",
      "preflightOption": Array [],
      "renderMode": "local",
    },
    Object {
      "format": "pdf",
      "path": "__WORKSPACE__/tests/fixtures/config/output2.pdf",
      "preflight": "press-ready",
      "preflightOption": Array [],
      "renderMode": "local",
    },
  ],
  "proxy": undefined,
  "quick": false,
  "readingProgression": "rtl",
  "rootUrl": "http://0.0.0.0:9876",
  "sandbox": false,
  "server": Object {
    "allowedHosts": Array [],
    "host": true,
    "port": 9876,
    "proxy": Object {
      "/api": "http://localhost:6789",
    },
  },
  "serverRootDir": "__WORKSPACE__/tests/fixtures/config",
  "singleDoc": false,
  "size": Object {
    "format": "size",
  },
  "static": Object {
    "/": Array [
      "root1",
      "root2",
    ],
    "/static": Array [
      "path/to/static",
    ],
  },
  "temporaryFilePrefix": "__TEMPORARY_FILE_PREFIX__",
  "themeIndexes": Set {
    Object {
      "importPath": undefined,
      "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/themes/node_modules/debug-theme",
      "name": "debug-theme",
      "registry": false,
      "specifier": "__WORKSPACE__/tests/fixtures/themes/debug-theme",
      "type": "package",
    },
    Object {
      "location": "__WORKSPACE__/tests/fixtures/config/workspaceDir/theme.css",
      "name": "theme.css",
      "source": "__WORKSPACE__/tests/fixtures/config/theme.css",
      "type": "file",
    },
  },
  "themesDir": "__WORKSPACE__/tests/fixtures/config/workspaceDir/themes",
  "timeout": 1,
  "title": "title",
  "vfmOptions": Object {
    "disableFormatHtml": true,
    "hardLineBreaks": true,
  },
  "viewer": undefined,
  "viewerInput": Object {
    "manifestPath": "__WORKSPACE__/tests/fixtures/config/workspaceDir/publication.json",
    "needToGenerateManifest": true,
    "type": "webpub",
  },
  "viewerParam": "foo=bar",
  "viteConfig": undefined,
  "viteConfigFile": true,
  "workspaceDir": "__WORKSPACE__/tests/fixtures/config/workspaceDir",
}
`;
